rendering-and-Data-fetching.md

```diff
@@ -1 +1,254 @@
+
+- Static Generation is the pre-rendering method that generates the HTML at build time, and reused Html page  on each request.
+- Server-side Rendering is the pre-rendering method that generates the HTML on each request.
+ 
+ ![static-generation](https://user-images.githubusercontent.com/7718220/89292420-06106480-d665-11ea-96f0-f65b460fc4ef.png)
+
+
+![server-side-rendering](https://user-images.githubusercontent.com/7718220/89292476-16284400-d665-11ea-9de4-9654c2ce7c57.png)
+
+
+# When to Use Static Generation VS Server-side Rendering
+
+Next js recommend using Static Generation (with and without data , we will take about it later)  which makes it much faster than having a server render the page on every request.
+#### Static Generation used with : 
+    - Static html pages
+    - the pages that contains a data that doesn't change frequently
+
+#### You can use Static Generation for many types of pages, including:
+
+    - Marketing pages
+    - Blog posts
+    - Help and documentation
+    
+On the other hand, Server-side Rendering is a good idea if your a page shows frequently updated data , and the page content changes on every request,
+In that case, you can use Server-Side Rendering. It will be slower, but the pre-rendered page will always be up-to-date.
+
+# Working with Data Fetching
+
+Data Ferching related with Pre-rendering , we talked about Pre-rendering (Static Generation and Server-side Rendering) , and we’ll talk in depths about data fetching strategies for each case .
+We’ll talk about the three unique Next.js functions you can use to fetch data for pre-rendering:
+
+ - #### `getStaticProps` : using with Static Generation to Fetch data at build time.
+ - #### `getStaticPaths` : using with Static Generation in dynamic routes to define a list of paths that have to be rendered to HTML at build time.
+ - #### `getServerSideProps` : using with Server-side Rendering to  Fetch data on each request.
+
+ ###### we will take in depths about this unique Next.js functions
+
+# Static Generation types
+
+![static-generation-without-data](https://user-images.githubusercontent.com/7718220/89324424-ee9aa100-d68f-11ea-9aab-764b93d3783f.png)
+
+
+![static-generation-with-data](https://user-images.githubusercontent.com/7718220/89324469-0114da80-d690-11ea-8aaf-570ee06937b0.png)
+
+# `getStaticProps` Function
+
+#### `getStaticProps` tell Next.js: “Hey, this page has some data dependencies — so when you pre-render this page at build time, make sure to resolve them first!”
+
+to run `getStaticProps` we must export an `async function` called `getStaticProps` from a page, the data return from functions will uaing as props by Next.js in pre-render of page at build time.
+
+In the following example we will export `async function getStaticProps` to fetch all post from [json place holder](https://jsonplaceholder.typicode.com/posts).
+the Function take `context` parameter and return `posts` as props to using in component to build post List 
```

Comment:
the Function takes* `context` parameter and returns* `posts` as props to use* them in 
 the* component to build post List